server/src/main/java/org/elasticsearch/cluster/coordination/DiscoveryUpgradeService.java

```diff
@@ -160,15 +164,22 @@ void countDown() {
 
     private class JoiningRound {
         private final boolean upgrading;
+        private final int minimumMasterNodes;
 
-        JoiningRound(boolean upgrading) {
+        JoiningRound(boolean upgrading, int minimumMasterNodes) {
             this.upgrading = upgrading;
+            this.minimumMasterNodes = minimumMasterNodes;
         }
 
         private boolean isRunning() {
             return joiningRound == this && isBootstrappedSupplier.getAsBoolean() == false;
         }
 
+        private boolean canBootstrap(Set<DiscoveryNode> discoveryNodes) {
+            return upgrading && 1 <= minimumMasterNodes
```

Comment:
The lenience worries me. Maybe we should require it to be `>0` if trying to join a 6.x master to block the "improperly configured" situation earlier?

Reply:
> Maybe we should require it to be >0 if trying to join a 6.x master to block the "improperly configured" situation earlier?

That would still not give us any hard guarantees that the value would be >0, as join validation will only help if the Zen1 node is a (recovered) master, as the join validation will otherwise not provide what we need. That situation will of course not correspond to a typical rolling restart anymore.

As we're using the 3 second pinging here, I'm not sure this brings us to a worse situation than what we have with Zen1 today. My suggestion is to relax the condition here, and we can then still make it stricter with extra validations in a follow-up if we see the need for it.

Reply:
Ok. This check also broke BWC (6.7 doesn't expose minimum_master_nodes yet which we treat as `-1`) so it's gone in 6e7a4f8da57.